File: software/rpi/libs/camera/ball_detection.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <algorithm>
#include <sstream>
#include <iomanip>

// Define struct to represent an IR point
struct IRPoint {
    cv::Point position;
    double intensity;
    double angle;
    double distance;
};

// Constants
const bool DEBUG = true;
const bool SAVE_POINTS = false;
const int MIN_CONTOUR_AREA = 10;
const int MIN_BRIGHTNESS = 130;
const cv::Point CENTRE_POINT(640 / 2 - 5, 480 / 2 + 30);

// Purple HSV range
const cv::Scalar PURPLE_LOWER(130, 50, 100);  // Expanded range to capture more variations
const cv::Scalar PURPLE_UPPER(175, 255, 255);

// Utility functions
double calculateDistance(const cv::Point& p1, const cv::Point& p2) {
    return std::sqrt(std::pow(p2.x - p1.x, 2) + std::pow(p2.y - p1.y, 2));
}

double calculateAngle(const cv::Point& point, const cv::Point& center = CENTRE_POINT) {
    double deltaX = point.x - center.x;
    double deltaY = point.y - center.y;
    double angle = std::atan2(deltaY, deltaX);
    return angle + M_PI / 2;
}

// Format floating point number to string with precision
std::string formatFloat(double value, int precision) {
    std::ostringstream ss;
    ss << std::fixed << std::setprecision(precision) << value;
    return ss.str();
}

// Detect bright purple (IR) points in the frame
std::pair<std::vector<IRPoint>, cv::Mat> detectIRPoints(const cv::Mat& frame) {
    // Convert to HSV colorspace
    cv::Mat hsv;
    cv::cvtColor(frame, hsv, cv::COLOR_BGR2HSV);
    
    // Create mask for purple color
    cv::Mat purpleMask;
    cv::inRange(hsv, PURPLE_LOWER, PURPLE_UPPER, purpleMask);
    
    // Additionally check for brightness in original frame
    cv::Mat gray, brightMask;
    cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
    cv::threshold(gray, brightMask, MIN_BRIGHTNESS, 255, cv::THRESH_BINARY);
    
    // Combine masks - we want purple AND bright
    cv::Mat irMask;
    cv::bitwise_and(purpleMask, brightMask, irMask);
    
    // Apply minimal morphology to clean up the mask
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
    cv::morphologyEx(irMask, irMask, cv::MORPH_OPEN, kernel);
    
    // Find contours in the mask
    std::vector<std::vector<cv::Point>> contours;
    std::vector<cv::Vec4i> hierarchy;
    cv::findContours(irMask.clone(), contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    
    // Filter contours by minimum area
    std::vector<IRPoint> irPoints;
    for (const auto& contour : contours) {
        double area = cv::contourArea(contour);
        if (area >= MIN_CONTOUR_AREA) {
            // Calculate centroid of the contour
            cv::Moments M = cv::moments(contour);
            if (M.m00 > 0) {
                int cx = static_cast<int>(M.m10 / M.m00);
                int cy = static_cast<int>(M.m01 / M.m00);
                cv::Point position(cx, cy);
                
                // Create and store point data
                IRPoint point;
                point.position = position;
                point.intensity = area;
                point.angle = calculateAngle(position) * 180.0 / M_PI;
                point.distance = calculateDistance(position, CENTRE_POINT);
                
                irPoints.push_back(point);
            }
        }
    }
    
    return {irPoints, irMask};
}

// Find the strongest IR source based on intensity
cv::Point findStrongestIRSource(const std::vector<IRPoint>& irPoints) {
    if (irPoints.empty()) {
        return cv::Point(-1, -1); // Invalid point
    }
    
    // Find maximum intensity point
    auto maxElement = std::max_element(irPoints.begin(), irPoints.end(),
        [](const IRPoint& a, const IRPoint& b) {
            return a.intensity < b.intensity;
        });
    
    return maxElement->position;
}

int main() {
    // Open video file
    std::string videoPath = "vid2.mp4";
    
    // Check if file exists (C++ way)
    std::ifstream f(videoPath.c_str());
    std::cout << "file exists? " << (f.good() ? "true" : "false") << std::endl;
    f.close();
    
    cv::VideoCapture cap(videoPath);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open video file." << std::endl;
        return -1;
    }
    
    // Statistics
    int allFramesTotalPoints = 0;
    int frameCount = 0;
    
    // Main loop
    while (cap.isOpened()) {
        cv::Mat frame;
        if (!cap.read(frame)) {
            break;
        }
        
        frameCount++;
        auto startTime = std::chrono::high_resolution_clock::now();
        cv::Mat output = frame.clone();
        
        // Mark center point
        cv::circle(output, CENTRE_POINT, 5, cv::Scalar(255, 255, 255), -1);
        
        // Detect IR points
        auto [currentFramePoints, irMask] = detectIRPoints(frame);
        
        // Update statistics
        allFramesTotalPoints += currentFramePoints.size();
        
        // Draw all detected IR points
        for (const auto& point : currentFramePoints) {
            // Size of circle based on intensity
            int radius = std::max(5, std::min(20, static_cast<int>(point.intensity / 10)));
            
            // Draw circle
            cv::circle(output, point.position, radius, cv::Scalar(255, 0, 255), -1);
            
            // Draw line from center to point
            cv::line(output, CENTRE_POINT, point.position, cv::Scalar(255, 0, 255), 2);
            
            // Display information
            std::string posText = "(" + std::to_string(point.position.x) + "," + 
                                  std::to_string(point.position.y) + ")";
            cv::putText(
                output, posText,
                cv::Point(point.position.x - 40, point.position.y - 15),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 0, 255), 1
            );
            
            std::string angleText = "Angle: " + formatFloat(point.angle, 1) + "°";
            cv::putText(
                output, angleText,
                cv::Point(point.position.x - 40, point.position.y + 15),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 0, 255), 1
            );
            
            std::string distText = "Dist: " + formatFloat(point.distance, 1);
            cv::putText(
                output, distText,
                cv::Point(point.position.x - 40, point.position.y + 30),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 0, 255), 1
            );
        }
        
        // Find and highlight the strongest IR source
        cv::Point strongestPoint = findStrongestIRSource(currentFramePoints);
        if (strongestPoint.x >= 0) {
            cv::circle(output, strongestPoint, 15, cv::Scalar(0, 255, 255), 2);
            cv::putText(
                output, "Strongest IR",
                cv::Point(strongestPoint.x - 40, strongestPoint.y - 30),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 255), 1
            );
        }
        
        // Debug visualization
        if (DEBUG) {
            // Create mask visualization
            cv::Mat irDisplay;
            cv::cvtColor(irMask, irDisplay, cv::COLOR_GRAY2BGR);
            cv::putText(
                irDisplay, "IR Mask",
                cv::Point(10, 30),
                cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2
            );
            
            // Information display
            cv::Mat infoDisplay = cv::Mat::zeros(frame.size(), frame.type());
            int detectedCount = currentFramePoints.size();
            
            cv::putText(
                infoDisplay,
                "Points in current frame: " + std::to_string(detectedCount),
                cv::Point(10, 30),
                cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2
            );
            
            cv::putText(
                infoDisplay,
                "Current frame: " + std::to_string(frameCount),
                cv::Point(10, 150),
                cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2
            );
            
            // Find the point data for the strongest point
            if (strongestPoint.x >= 0) {
                auto strongestData = std::find_if(
                    currentFramePoints.begin(), currentFramePoints.end(),
                    [&strongestPoint](const IRPoint& p) {
                        return p.position.x == strongestPoint.x && p.position.y == strongestPoint.y;
                    }
                );
                
                if (strongestData != currentFramePoints.end()) {
                    cv::putText(
                        infoDisplay,
                        "Strongest IR Angle: " + formatFloat(strongestData->angle, 1) + "°",
                        cv::Point(10, 70),
                        cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2
                    );
                    
                    cv::putText(
                        infoDisplay,
                        "Strongest IR Distance: " + formatFloat(strongestData->distance, 1),
                        cv::Point(10, 110),
                        cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2
                    );
                }
            }
            
            // Combine displays
            int h = frame.rows, w = frame.cols;
            cv::Mat irDisplayResized, infoDisplayResized;
            cv::resize(irDisplay, irDisplayResized, cv::Size(w / 2, h / 2));
            cv::resize(infoDisplay, infoDisplayResized, cv::Size(w / 2, h / 2));
            
            cv::Mat debugTop;
            cv::hconcat(irDisplayResized, infoDisplayResized, debugTop);
            
            cv::Mat outputResized;
            cv::resize(output, outputResized, cv::Size(w, h / 2));
            
            cv::Mat debugView;
            cv::vconcat(debugTop, outputResized, debugView);
            
            cv::imshow("IR Detection Debug", debugView);
        }
        
        cv::imshow("IR Detection", output);
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        
        std::cout << "Frame " << frameCount << ": Found " << currentFramePoints.size() 
                  << " points in " << duration.count() << "ms" << std::endl;
        
        // Print points in current frame
        if (!currentFramePoints.empty()) {
            std::cout << "Points in frame " << frameCount << ":" << std::endl;
            for (size_t i = 0; i < currentFramePoints.size(); i++) {
                const auto& point = currentFramePoints[i];
                std::cout << "  Point " << (i+1) << ": Position (" 
                          << point.position.x << "," << point.position.y 
                          << "), Angle: " << formatFloat(point.angle, 1) 
                          << "°, Distance: " << formatFloat(point.distance, 1) << std::endl;
            }
        }
        
        int key = cv::waitKey(25);
        if (key == 'q') {
            break;
        }
    }
    
    // Print summary
    std::cout << "Total frames processed: " << frameCount << std::endl;
    std::cout << "Total points detected across all frames: " << allFramesTotalPoints << std::endl;
    
    cap.release();
    cv::destroyAllWindows();
    
    // In C++, we would return the points through a function or use a global variable
    // if this were part of a larger program, but for a standalone script we print them
    std::cout << "All points in the last frame are stored in memory" << std::endl;
    
    return 0;
}

File: software/rpi-direct/libs/camera/scripts/ball_detection.py
import cv2
import numpy as np
import os
import datetime
import math
import time

DEBUG = True
SAVE_POINTS = False  # Changed to False to disable CSV saving

video_path = "vid2.mp4"
print("file exists? " + str(os.path.exists("vid2.mp4")))
cap = cv2.VideoCapture(video_path)

# Constants for IR detection
MIN_CONTOUR_AREA = 10  # Much smaller than goalpost since we're looking for points
MIN_BRIGHTNESS = 130  # Lowered threshold for detecting bright spots
CENTRE_POINT = [640 // 2 - 5, 480 // 2 + 30]

# Adjusted Purple HSV range
PURPLE_LOWER = np.array([130, 50, 100])  # Expanded range to capture more variations
PURPLE_UPPER = np.array([175, 255, 255])


def calculate_distance(p1, p2):
    """Calculate Euclidean distance between two points."""
    return np.sqrt((p2[0] - p1[0]) ** 2 + (p2[1] - p1[1]) ** 2)


def calculate_angle(point, center=CENTRE_POINT):
    """Calculate angle from center to point (in radians)."""
    delta_x = point[0] - center[0]
    delta_y = point[1] - center[1]
    angle = math.atan2(delta_y, delta_x)
    return angle + math.pi / 2


def detect_ir_points(frame):
    """
    Detect bright purple (IR) points in the frame.
    Returns a list of all purple points in the current frame and the mask.
    """
    # Convert to HSV colorspace
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    # Create mask for purple color
    purple_mask = cv2.inRange(hsv, PURPLE_LOWER, PURPLE_UPPER)

    # Additionally check for brightness in original frame
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    bright_mask = cv2.threshold(gray, MIN_BRIGHTNESS, 255, cv2.THRESH_BINARY)[1]

    # Combine masks - we want purple AND bright
    ir_mask = cv2.bitwise_and(purple_mask, bright_mask)

    # Apply minimal morphology to clean up the mask
    kernel = np.ones((3, 3), np.uint8)
    ir_mask = cv2.morphologyEx(ir_mask, cv2.MORPH_OPEN, kernel)

    # Find contours in the mask
    contours, _ = cv2.findContours(ir_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Filter contours by minimum area
    ir_points = []
    for contour in contours:
        area = cv2.contourArea(contour)
        if area >= MIN_CONTOUR_AREA:
            # Calculate centroid of the contour
            M = cv2.moments(contour)
            if M["m00"] > 0:
                cx = int(M["m10"] / M["m00"])
                cy = int(M["m01"] / M["m00"])
                ir_points.append(
                    {
                        "position": (cx, cy),
                        "intensity": area,
                        "angle": calculate_angle((cx, cy)) * 180 / math.pi,
                        "distance": calculate_distance((cx, cy), CENTRE_POINT),
                    }
                )

    return ir_points, ir_mask


def find_strongest_ir_source(ir_points):
    """Find the strongest IR source based on intensity."""
    if not ir_points:
        return None

    # Sort by intensity (descending)
    sorted_points = sorted(ir_points, key=lambda p: p["intensity"], reverse=True)
    return sorted_points[0]["position"]


# For statistics - can be removed if not needed
all_frames_total_points = 0
frame_count = 0

# Main loop
while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    frame_count += 1
    start_time = datetime.datetime.now()
    output = frame.copy()

    # Mark center point
    cv2.circle(output, tuple(CENTRE_POINT), 5, (255, 255, 255), -1)

    # Detect IR points - this returns all purple points in the current frame
    current_frame_points, ir_mask = detect_ir_points(frame)

    # Update statistics
    all_frames_total_points += len(current_frame_points)

    # Draw all detected IR points in current frame
    for point in current_frame_points:
        position = point["position"]
        intensity = point["intensity"]
        angle_degrees = point["angle"]
        distance = point["distance"]

        # Size of circle based on intensity
        radius = max(5, min(20, int(intensity / 10)))

        # Draw circle
        cv2.circle(output, position, radius, (255, 0, 255), -1)

        # Draw line from center to point
        cv2.line(output, tuple(CENTRE_POINT), position, (255, 0, 255), 2)

        # Display information
        cv2.putText(
            output,
            f"({position[0]},{position[1]})",
            (position[0] - 40, position[1] - 15),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (255, 0, 255),
            1,
        )
        cv2.putText(
            output,
            f"Angle: {angle_degrees:.1f}°",
            (position[0] - 40, position[1] + 15),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (255, 0, 255),
            1,
        )
        cv2.putText(
            output,
            f"Dist: {distance:.1f}",
            (position[0] - 40, position[1] + 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (255, 0, 255),
            1,
        )

    # Find and highlight the strongest IR source
    strongest_point = find_strongest_ir_source(current_frame_points)
    if strongest_point:
        cv2.circle(output, strongest_point, 15, (0, 255, 255), 2)
        cv2.putText(
            output,
            "Strongest IR",
            (strongest_point[0] - 40, strongest_point[1] - 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.5,
            (0, 255, 255),
            1,
        )

    # Debug visualization
    if DEBUG:
        # Create mask visualization
        ir_display = cv2.cvtColor(ir_mask, cv2.COLOR_GRAY2BGR)
        cv2.putText(
            ir_display,
            "IR Mask",
            (10, 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            (0, 255, 0),
            2,
        )

        # Information display
        info_display = np.zeros_like(frame)
        detected_count = len(current_frame_points)

        cv2.putText(
            info_display,
            f"Points in current frame: {detected_count}",
            (10, 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            (0, 255, 0),
            2,
        )

        cv2.putText(
            info_display,
            f"Current frame: {frame_count}",
            (10, 150),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            (0, 255, 0),
            2,
        )

        if strongest_point:
            # Find the point data for the strongest point
            strongest_data = next(
                (p for p in current_frame_points if p["position"] == strongest_point),
                None,
            )

            if strongest_data:
                cv2.putText(
                    info_display,
                    f"Strongest IR Angle: {strongest_data['angle']:.1f}°",
                    (10, 70),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.7,
                    (0, 255, 0),
                    2,
                )
                cv2.putText(
                    info_display,
                    f"Strongest IR Distance: {strongest_data['distance']:.1f}",
                    (10, 110),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.7,
                    (0, 255, 0),
                    2,
                )

        # Combine displays
        h, w = frame.shape[:2]
        ir_display = cv2.resize(ir_display, (w // 2, h // 2))
        info_display = cv2.resize(info_display, (w // 2, h // 2))

        debug_top = np.hstack([ir_display, info_display])
        debug_view = np.vstack([debug_top, cv2.resize(output, (w, h // 2))])
        cv2.imshow("IR Detection Debug", debug_view)

    cv2.imshow("IR Detection", output)
    end_time = datetime.datetime.now()
    print(
        f"Frame {frame_count}: Found {len(current_frame_points)} points in {end_time - start_time}"
    )

    # Print points in current frame
    if current_frame_points:
        print(f"Points in frame {frame_count}:")
        for i, point in enumerate(current_frame_points):
            print(
                f"  Point {i+1}: Position {point['position']}, Angle: {point['angle']:.1f}°, Distance: {point['distance']:.1f}"
            )

    if cv2.waitKey(25) & 0xFF == ord("q"):
        break

# Print summary
print(f"Total frames processed: {frame_count}")
print(f"Total points detected across all frames: {all_frames_total_points}")

cap.release()
cv2.destroyAllWindows()

# This makes the current frame's points available
print(
    f"All {len(current_frame_points)} points in the last frame are stored in 'current_frame_points' variable"
)
detected_points = current_frame_points  # Store in a variable that can be used later
